core/ir/synthetic/encode.c

```diff
@@ -41,92 +41,118 @@
 
 /* Encodes DR instruction representation \p instr into raw bytes \p encoded_instr.
  * Returns next instruction's pc.
- * A description of the encoding scheme is provided in core/ir/synthetic/decode.c.
- * We assume all encoded values to be little-endian.
+ * The encoding scheme followed is described in core/ir/synthetic/encoding_common.h.
  */
 byte *
 encode_to_synth(dcontext_t *dcontext, instr_t *instr, byte *encoded_instr)
 {
-    uint encoding = 0;
-    uint shift = 0;
+    /* Interpret the first 4 bytes of encoded_instr (which are always present) as a uint
+     * for easier retrieving of category, eflags, #src, and #dst values.
+     * We can do this safely because encoded_instr is 4 bytes aligned.
+     */
+    uint *encoding = ((uint *)&encoded_instr[0]);
+    *encoding = 0;
 
     /* Encode number of destination operands.
+     * Note that a destination operand that is a memory renference, should have its
+     * registers (if any) counted as source operands, since they are being read.
+     * We use used_[src|dst]_reg_map to keep track of registers we've seen and avoid
+     * duplicates.
      */
-    uint original_num_dsts = (uint)instr_num_dsts(instr);
+    byte used_src_reg_map[MAX_NUM_REGS];
```

Comment:
bool instead of byte with true and false instead of 0 and 1 is nicer?

Reply:
Done.